### PR TITLE
feat: make `to` address optional for transaction data

### PR DESCRIPTION
**Pull request by IAvecilla**
## What ❔

This PR is the first part of [this one](https://github.com/matter-labs/zksync-era/pull/2631). 

We decided to separate it into smaller PRs to make it easier to read and review.

This PR contains only the changes to the `TransactionData` structs for the different VM versions, modifying them for an optional field.

The other parts are currently available in our repo fork, as they are based on this branch. The other PRs are:
- [feat: evm simulator as base system contract](https://github.com/lambdaclass/zksync-era/pull/256): integrates the EVM interpreter contract as a base system contract and includes all necessary updates for zksync-era to run with it.
- [feat: make evm simulator optional for the server](https://github.com/lambdaclass/zksync-era/pull/257): makes the evm simulator optional to run using a config file.

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: core/lib/dal/src/transactions_dal.rs
Comment (0xVolosnikov) [inline core/lib/dal/src/transactions_dal.rs:63]:
```suggestion
        let contract_address_as_bytes: &[u8] = if contract_address.is_none() {
```

File: core/lib/dal/src/transactions_dal.rs
Comment (0xVolosnikov) [inline core/lib/dal/src/transactions_dal.rs:172]:
```suggestion
        let contract_address_as_bytes: &[u8] = if contract_address.is_none() {
```

File: core/lib/dal/src/transactions_dal.rs
Comment (0xVolosnikov) [inline core/lib/dal/src/transactions_dal.rs:821]: Nit: do we need this change?

File: core/lib/types/src/l2/mod.rs
Comment (0xVolosnikov) [inline core/lib/types/src/l2/mod.rs:235]: Shouldn't it be also `-> Option<Address>`?

File: prover/crates/lib/keystore/src/utils.rs
Comment (0xVolosnikov) [inline prover/crates/lib/keystore/src/utils.rs:118]: Nit: unnecessary diff

File: core/lib/multivm/src/versions/vm_1_3_2/transaction_data.rs
Comment (IAvecilla) [reply core/lib/multivm/src/versions/vm_1_3_2/transaction_data.rs:173]: You mean to apply this change for older VMs? At least for `vm_latest`, we can have a `None` at this point, so we need to take the default `U256` to avoid panicking.

Comment (0xVolosnikov) [reply core/lib/multivm/src/versions/vm_1_3_2/transaction_data.rs:173]: > This method is to be used only in tests, when we want to bypass the checks imposed

I don't think the `None` scenario is implied in tests. On the other hand, I don't think it's a problem.

File: prover/crates/lib/keystore/src/utils.rs
Comment (IAvecilla) [reply prover/crates/lib/keystore/src/utils.rs:118]: I think that `zk fmt` is adding it.

File: core/lib/dal/src/transactions_dal.rs
Comment (IAvecilla) [reply core/lib/dal/src/transactions_dal.rs:821]: I also updated the implementation for obtaining the `contract_address_bytes` to reduce the boilerplate.

File: core/lib/types/src/l2/mod.rs
Comment (IAvecilla) [reply core/lib/types/src/l2/mod.rs]: Indeed, updated!

File: core/lib/multivm/src/versions/vm_1_3_2/transaction_data.rs
Comment (IAvecilla) [reply core/lib/multivm/src/versions/vm_1_3_2/transaction_data.rs:173]: Correct me if I'm wrong, but this function `abi_encode_with_custom_factory_deps` is used every time we send a new transaction since it's called by the `BootloaderTx::new()` function, which happens every time we push a new transaction into the bootloader memory.

Changing this line to an `unwrap()` in `vm_latest` causes a panic every time I try to deploy an EVM contract. 
Maybe I'm missing something here, so please let me know.

**Comment by 0xVolosnikov** [comment]
Lgtm, waiting for code owners to review

File: core/lib/types/src/tx/execute.rs
Comment (perekopskiy) [inline core/lib/types/src/tx/execute.rs:78]: is it correct way to do it? AFAIU it means that signatures for txs with `tx.contract_address = null` and `tx. contract_address = Address::zero()` will be the same which doesn't look correct

File: core/lib/types/src/transaction_request.rs
Comment (perekopskiy) [inline core/lib/types/src/transaction_request.rs:821]: so this means we start allowing txs with to=null in our system but process them as if `to=0x00..00`. I would continue rejecting such txs until evm equivalence is on, otherwise it will be confusing for everyone

File: core/lib/dal/src/consensus/mod.rs
Comment (IAvecilla) [reply core/lib/dal/src/consensus/mod.rs]: Right, I changed it to get the actual `Option` instead of using `required`, is that the correct way to do it?

File: core/lib/dal/src/transactions_dal.rs
Comment (IAvecilla) [reply core/lib/dal/src/transactions_dal.rs]: Done! I had to add some type conversions in a few queries since that seems to be the correct way to handle Null inserts with `sqlx`

File: core/lib/types/src/tx/execute.rs
Comment (IAvecilla) [reply core/lib/types/src/tx/execute.rs]: I agree that it might be incorrect for the signature to remain the same when the address is null or `0x00`. I've changed it to only include the `to` field as a member when the contract address is `Some(..)`, and not include it when it's `None`. 

I'm not entirely sure if this is the best approach or if it could introduce additional issues. While reviewing other parts of the code, I noticed that when a member has an `Option` type, it usually defaults when it is `None`, like [here](https://github.com/matter-labs/zksync-era/blob/3f1d1a26076d5bf2fd65c31876d2a32a3cbfff0c/core/lib/types/src/transaction_request.rs#L339), but this might be a case where it's not possible to get `None` as a value. Let me know it there's a better way to do this

File: core/lib/types/src/transaction_request.rs
Comment (IAvecilla) [reply core/lib/types/src/transaction_request.rs:826]: Done! I added the check above and included a comment for its removal once the EVM interpreter is integrated.

File: core/lib/types/src/l1/mod.rs
Comment (jrchatruc) [reply core/lib/types/src/l1/mod.rs:277]: L1 transactions come from `L2TransactionRequest`s, which have to contain an address, so I don't think `contract_address` is nullable.

I imagine making it non-nullable through the type system would involve making a new `Execute` struct specific for L1 transactions where the `contract_address` field is not an `Option` and make L1Tx contain it instead of the regular `Execute`. Is that reasonable or do you think it's better to unwrap/use try_from to return an error instead?

File: core/lib/dal/src/consensus/mod.rs
Comment (perekopskiy) [reply core/lib/dal/src/consensus/mod.rs]: Yes! Please also change comment in .proto file (`optional bytes contract_address = 1; // required; H160` -> `optional bytes contract_address = 1; // optional; H160`

File: core/lib/types/src/l1/mod.rs
Comment (perekopskiy) [reply core/lib/types/src/l1/mod.rs:277]: > L1 transactions come from L2TransactionRequests, which have to contain an address, so I don't think contract_address is nullable.

Yes, that's the current state, I was asking more about future plans. I didn't expect a definite answer but I wanted to raise this topic

> Is that reasonable or do you think it's better to unwrap/use try_from to return an error instead?

I think it's reasonable long-term but not as a part of this PR. So let's unwrap/return error for now but keep it in mind, cc @0xVolosnikov

**Review by perekopskiy — APPROVED** [review]

**Pull request merged**
